### ISSUE TITLE
[Feature Request] Support for Latent Interoperability

### ISSUE THREAD

**Issue by iwr-redmond** [Teriks/dgenerate#41]
The [SD Latent Interposers](https://huggingface.co/city96/SD-Latent-Interposer/tree/main/v4.0) allow latents to be converted to other architecture types, eliminating the need for VAE encoding and decoding. This facilitates SDXL Refiner-style inference across [most architecture pairs](https://github.com/city96/SD-Latent-Interposer?tab=readme-ov-file#supported-models), with the main limitation being an inability to convert latents _to_ Flux or Stable Cascade.

The interposer inference code is [here](https://github.com/city96/SD-Latent-Interposer/blob/main/interposer.py).

Supporting latent interoperability may be a breaking change, as pipelines typically conclude with VAE encoding. While the code itself appears straightforward, I reckon that the changes to the CLI may be slightly more complex.

**Comment by Teriks** [comment]
output type `latent` for diffusers skips the decode.  Potentially could add an output format which just writes the tensor do disk as an intermediate, the difficult part is accepting this form of input as an “image”, as everything here is not setup for that

**Comment by iwr-redmond** [comment]
> the difficult part is accepting this form of input as an “image”, as everything here is not setup for that

I suspected something like that...

**Comment by Teriks** [comment]
Maybe the `mediainput` module can be made friendly to this sort of input, generally the latents can be passed into the pipeline in the same argument that a PIL image would be.

There are a lot of corner cases in that module that would need to be written so that the existing logic for animated inputs could work along side a raw latent that is to be treated as a static image generically.  Given that preprocessing / resizing on such an input does not make sense.  

If not then passing the latent in requires a new argument, given that latents can be given for all possible image inputs, would be better to modify media input possibly

And then some in the `pipelinewrapper` to make sure image processing operations are not attempted on the latents

Will have to investigate

**Comment by Teriks** [comment]
This does support one liners from the CLI, just the downside is more cyclomatic complexity in the code.

I am around 90 percent done with implementing latents as an interchange format minus the interposing, which I think is a good start here.

Mostly need to sort out input validation for the number of latent channels and image space dimensions etc. to avoid reporting nasty diffusers exceptions to the user, as well as the behavior of `—image-seed-strengths` when accepting latents.

i.e. latents can be passed into `—image-seeds` for image inputs (all forms except masks probably) in the form of a tensor file `pt`, `pth`, `safetensors`

Resizing and image processing operations on latents are ignored with warnings.

And output directly to latents with `—image-format` `pt`, `pth`, `safetensors`

Caveat being adetailer cannot support direct latents output due to requiring post processing in image space, theoretically the composited image could be re-encoded just to provide a way to output in latent space.

Not sure how necessary that is, unless the latents could be consumed by another application outside this one.

Another route would be to have a `—latents-input-format` and `—latents-output-format`,  and automatically convert upon output or while receiving input from `—image-seeds`

In the advanced case this supports chaining inference on the latents across more than just a secondary pipeline, just requires multiple tool calls and an interchange file.

`—sdxl-refiner` is a bit special as the SDXL img2img pipeline itself implements arguments for cooperative denoising which are used when supported, and otherwise edit mode is used (refine on decoded latents), it is sort of coupled very tightly to the the primary SDXL pipeline in a similar way to the Stable Cascade decoder

I definitely would like to avoid introducing more secondary passes in monolithic fashion, just due to cyclomatic complexity and the possible introduction of combinatoric arguments for the secondary pipeline

> Perhaps a less complex starting point would be to create `--sd-refiner`, `--sd3-refiner`, and `--kolors-refiner` arguments to compliment the existing `--sdxl-refiner` option, and use the interposers to transfer latents automatically from the source generation pipeline (Flux, Cascade, SD3x, SDXL, SD1x) to the refiner one (SD1x, SDXL, SD3x).

**Comment by iwr-redmond** [comment]
> Another route would be to have a —latents-input-format and —latents-output-format, and automatically convert upon output or while receiving input from —image-seeds

That sounds sensible.

**Comment by Teriks** [comment]
So far, latents can be written with ``--image-format (pt, pth, or safetensors)``, and read in place of images for img2img operations with ``--image-seeds``

``--image-seed-strengths`` affects generation in the same way as it would when passing in a normal image, i.e. the incoming latents
have noise added per the strength value

**Comment by Teriks** [comment]
Something like this for interposing with the comfy UI interposer.

I notice that when the interposed latents get decoded by the VAE, the quality is not that great, could be something to do
with the scaling factor / vae configs used by comfy.  I think this model is trained with those VAE configs in mind and diffusers
is a bit different.

```python

import torch
import dgenerate
import dgenerate.pipelinewrapper as _pipelinewrapper
import os
import torch.nn as nn
from safetensors.torch import load_file
from huggingface_hub import hf_hub_download

config = {
    "v1-to-xl": {"ch_in": 4, "ch_out": 4, "ch_mid": 64, "scale": 1.0, "blocks": 12},
    "v1-to-v3": {"ch_in": 4, "ch_out":16, "ch_mid": 64, "scale": 1.0, "blocks": 12},
    "xl-to-v1": {"ch_in": 4, "ch_out": 4, "ch_mid": 64, "scale": 1.0, "blocks": 12},
    "xl-to-v3": {"ch_in": 4, "ch_out":16, "ch_mid": 64, "scale": 1.0, "blocks": 12},
    "v3-to-v1": {"ch_in":16, "ch_out": 4, "ch_mid": 64, "scale": 1.0, "blocks": 12},
    "v3-to-xl": {"ch_in":16, "ch_out": 4, "ch_mid": 64, "scale": 1.0, "blocks": 12},
    "fx-to-v1": {"ch_in":16, "ch_out": 4, "ch_mid": 64, "scale": 1.0, "blocks": 12},
    "fx-to-xl": {"ch_in":16, "ch_out": 4, "ch_mid": 64, "scale": 1.0, "blocks": 12},
    "fx-to-v3": {"ch_in":16, "ch_out":16, "ch_mid": 64, "scale": 1.0, "blocks": 12},
    "ca-to-v1": {"ch_in": 4, "ch_out": 4, "ch_mid": 64, "scale": 0.5, "blocks": 12},
    "ca-to-xl": {"ch_in": 4, "ch_out": 4, "ch_mid": 64, "scale": 0.5, "blocks": 12},
    "ca-to-v3": {"ch_in": 4, "ch_out":16, "ch_mid": 64, "scale": 0.5, "blocks": 12},
}

class ResBlock(nn.Module):
    """Block with residuals"""
    def __init__(self, ch):
        super().__init__()
        self.join = nn.ReLU()
        self.norm = nn.BatchNorm2d(ch)
        self.long = nn.Sequential(
            nn.Conv2d(ch, ch, kernel_size=3, stride=1, padding=1),
            nn.SiLU(),
            nn.Conv2d(ch, ch, kernel_size=3, stride=1, padding=1),
            nn.SiLU(),
            nn.Conv2d(ch, ch, kernel_size=3, stride=1, padding=1),
            nn.Dropout(0.1)
        )
    def forward(self, x):
        x = self.norm(x)
        return self.join(self.long(x) + x)

class ExtractBlock(nn.Module):
    """Increase no. of channels by [out/in]"""
    def __init__(self, ch_in, ch_out):
        super().__init__()
        self.join  = nn.ReLU()
        self.short = nn.Conv2d(ch_in, ch_out, kernel_size=3, stride=1, padding=1)
        self.long  = nn.Sequential(
            nn.Conv2d( ch_in, ch_out, kernel_size=3, stride=1, padding=1),
            nn.SiLU(),
            nn.Conv2d(ch_out, ch_out, kernel_size=3, stride=1, padding=1),
            nn.SiLU(),
            nn.Conv2d(ch_out, ch_out, kernel_size=3, stride=1, padding=1),
            nn.Dropout(0.1)
        )
    def forward(self, x):
        return self.join(self.long(x) + self.short(x))

class InterposerModel(nn.Module):
    """
    NN layout, ported from:
    https://github.com/city96/SD-Latent-Interposer/blob/main/interposer.py
    """
    def __init__(self, ch_in=4, ch_out=4, ch_mid=64, scale=1.0, blocks=12):
        super().__init__()
        self.ch_in  = ch_in
        self.ch_out = ch_out
        self.ch_mid = ch_mid
        self.blocks = blocks
        self.scale  = scale

        self.head = ExtractBlock(self.ch_in, self.ch_mid)
        self.core = nn.Sequential(
            nn.Upsample(scale_factor=self.scale, mode="nearest"),
            *[ResBlock(self.ch_mid) for _ in range(blocks)],
            nn.BatchNorm2d(self.ch_mid),
            nn.SiLU(),
        )
        self.tail = nn.Conv2d(self.ch_mid, self.ch_out, kernel_size=3, stride=1, padding=1)

    def forward(self, x):
        y = self.head(x)
        z = self.core(y)
        return self.tail(z)

class ComfyLatentInterposer:
    """Custom node"""
    def __init__(self):
        self.version = 4.0 # network revision
        self.loaded = None # current model name
        self.model = None  # current model

    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "samples": ("LATENT", ),
                "latent_src": (["v1", "xl", "v3", "fx", "ca"],),
                "latent_dst": (["v1", "xl", "v3"],),
            }
        }

    RETURN_TYPES = ("LATENT",)
    FUNCTION     = "convert"
    CATEGORY     = "latent"
    TITLE        = "Latent Interposer"

    def get_model_path(self, model_name):
        fname = f"{model_name}_interposer-v{self.version}.safetensors"
        path  = os.path.join(os.path.dirname(os.path.realpath(__file__)),"models")

        # local path: [models/xl-to-v1_interposer-v4.2.safetensors]
        if os.path.isfile(os.path.join(path, fname)):
            print("LatentInterposer: Using local model")
            return os.path.join(path, fname)

        # local path: [models/v4.2/xl-to-v1_interposer-v4.2.safetensors]
        if os.path.isfile(os.path.join(path, os.path.join(f"v{self.version}", fname))):
            print("LatentInterposer: Using local model")
            return os.path.join(path, os.path.join(f"v{self.version}", fname))

        # huggingface hub fallback
        print("LatentInterposer: Using HF Hub model")
        return str(hf_hub_download(
            repo_id   = "city96/SD-Latent-Interposer",
            subfolder = f"v{self.version}",
            filename  = fname,
        ))

    def convert(self, samples, latent_src, latent_dst):
        samples = samples.copy()
        if latent_src == latent_dst:
            return (samples,)

        model_name = f"{latent_src}-to-{latent_dst}"
        if model_name not in config:
            raise ValueError(f"No model exists for this conversion! ({model_name})")

        # only reload if changed
        if self.loaded != model_name or self.model is None:
            # load/init model
            path = self.get_model_path(model_name)
            model = InterposerModel(**config[model_name])
            model.eval()
            model.load_state_dict(load_file(path))
            # keep for later runs
            self.model = model
            self.loaded = model_name

        lt = samples["samples"]
        with torch.no_grad():
            # force FP32, always run on CPU
            lt = self.model(
                lt.cpu().float()
            ).to(lt.device).to(lt.dtype)
        samples["samples"] = lt
        return (samples,)


sd_pipe = _pipelinewrapper.DiffusionPipelineWrapper(
    'stable-diffusion-v1-5/stable-diffusion-v1-5',
    model_type=_pipelinewrapper.ModelType.TORCH
)

sdxl_pipe = _pipelinewrapper.DiffusionPipelineWrapper(
    'stabilityai/stable-diffusion-xl-base-1.0',
    model_type=_pipelinewrapper.ModelType.TORCH_SDXL,
    dtype=_pipelinewrapper.DataType.FLOAT16,
    variant='fp16'
)

interposer = ComfyLatentInterposer()

args = _pipelinewrapper.DiffusionArguments()

args.seed = 0
args.prompt = dgenerate.Prompt.parse('A beautiful landscape with mountains and a river')

# SD1.5 pipe should output latents
args.output_latents = True

args.width = 1024
args.height = 1024

# Generate an image using SD1.5
sd_result = sd_pipe(args)

# Add a batch dimension to the latents
latents_for_interposer = sd_result.latent.unsqueeze(0)

# Interposer expects unscaled latents
latents_for_interposer = latents_for_interposer / 0.18215

# SDXL pipe should output an image
args.output_latents = False

# 90 percent noise addition
args.image_seed_strength = 0.9

# Convert the latents using the interposer
interposed_latents = interposer.convert({'samples': latents_for_interposer}, latent_src='v1', latent_dst='xl')[0]['samples']

# pipeline expects scaled latents
interposed_latents = interposed_latents * 0.18215

args.images = [interposed_latents]

# Generate SDXL image with the interposed latents as img2img input
sdxl_result = sdxl_pipe(args)

# Decode the input latents with the SDXL VAE, this should give an idea
# of what the "image" looked like before we used it for an img2img operation
sdxl_pipe.decode_latents(args.images[0], args)[0].save('sdxl_vae_decoded.png')

# Save the final SDXL output image
sdxl_result.image.save('sdxl_output.png')

```

**Comment by Teriks** [comment]
> Take a look through the linked repository, as the [configs](https://github.com/city96/SD-Latent-Interposer/tree/main/config) and [VAE classes](https://github.com/city96/SD-Latent-Interposer/blob/main/vae.py) may be helpful, with the latter relying on [Diffusers](https://github.com/city96/SD-Latent-Interposer/blob/9344ddf53c3f4164a753f1ab30885ad99941a437/vae.py#L2) rather than Comfy.

scale = 1/8 is probably better in the example than 0.18215

**Comment by Teriks** [comment]
I think this is good to go, it should be easier to add different latent space manipulations in the future with the new plugin type
